field_RANKSUFF_module.fypp

```diff
@@ -342,7 +356,86 @@ CONTAINS
 
   CONTAINS
 
-#:for d in range (0, ft.rank+1)
+#:for d1 in range (0, ft.rank)
+#:for d2 in range (d1+1, ft.rank+1)
+    SUBROUTINE COPY_2D_DIM${d1}$_${d2}$_CONTIGUOUS (HST, DEV, KDIR, QUEUE)
+#ifdef _OPENACC
+    USE OPENACC
+    USE CUDAFOR
+#endif
+    ${ft.type}$, POINTER :: HST (${ft.shape}$), DEV (${ft.shape}$)
+    INTEGER (KIND=JPIM),           INTENT (IN) :: KDIR
+    INTEGER (KIND=JPIM), OPTIONAL, INTENT (IN) :: QUEUE
+#ifdef _OPENACC
+    INTEGER (KIND=JPIM) :: IHST_PITCH, IDEV_PITCH, IRET
+    INTEGER (KIND=JPIM) :: IWIDTH, IHEIGHT, ISHP(${ft.rank+1}$)
+  #:if d2 < ft.rank
+    INTEGER :: ${', '.join (list (map (lambda i: 'J' + str (i+1), range (d2, ft.rank))))}$
+  #:endif
+    INTEGER(KIND=CUDA_STREAM_KIND) :: STREAM
+
+    ISHP(1) = 1
+    ISHP(2:) = SHAPE(HST)
+    IWIDTH = PRODUCT(ISHP(1:${d1+1}$))
+    IHEIGHT = PRODUCT(ISHP(${d1+2}$:${d2+1}$))
+
+#:set lb = lambda arr, i: f'LBOUND({arr}, {i+1})'
+#:set lbnds = lambda arr, start, end: [lb(arr, i) for i in range(start, end)]
+#:set this_slice = lambda arr: ', '.join(lbnds(arr, 0, ft.rank))
+#:set next_slice = lambda arr: ', '.join(lbnds(arr, 0, d1) + [lb(arr, d1)+'+1'] + lbnds(arr, d1+1, ft.rank))
+    IHST_PITCH = (LOC (HST(${next_slice('HST')}$)) - LOC (HST (${this_slice('HST')}$)))/KIND(HST)
+    IDEV_PITCH = (LOC (DEV(${next_slice('DEV')}$)) - LOC (DEV (${this_slice('DEV')}$)))/KIND(DEV)
+
+  #:for e in range (ft.rank, d2, -1)
+    ${'  ' * (ft.rank - e)}$DO J${e}$ = LBOUND (HST, ${e}$), UBOUND (HST, ${e}$)
+  #:endfor
+  #:set indent = '  ' * (ft.rank - d2 - 1)
+  #:set ar = lambda arr: ', '.join(lbnds(arr, 0, d2) + [f'J{i+1}' for i in range(d2, ft.rank)])
+    ${indent}$  IF (KDIR == NH2D) THEN
+    ${indent}$    !$acc host_data use_device (DEV)
+    ${indent}$    IF(PRESENT(QUEUE)) THEN
+    ${indent}$      CALL ACC_SET_CUDA_STREAM(QUEUE, STREAM)
+    ${indent}$      IRET = CUDAMEMCPY2DASYNC (DEV (${ar('DEV')}$), IDEV_PITCH, &
+    ${indent}$                              & HST (${ar('HST')}$), IHST_PITCH, &
+    ${indent}$                              & IWIDTH, IHEIGHT, CUDAMEMCPYHOSTTODEVICE, &
+    ${indent}$                              & STREAM)
+    ${indent}$    ELSE
+    ${indent}$      IRET = CUDAMEMCPY2D (DEV (${ar('DEV')}$), IDEV_PITCH, &
```

Comment:
Two questions : 
- CUDAMEMCPY2DASYNC seems to require a CUDAMEMCPYHOSTTODEVICE/CUDAMEMCPYDEVICETOHOST argument, while CUDAMEMCPY2D does not; is it normal ?
- the direction of transfers triggered by CUDAMEMCPY2DASYNC/CUDAMEMCPY2D is influenced by the order of HST/DEV : H2D with DEV as first argument, D2H with HST as first argument. Correct ?

Reply:
- You are right, I can do this better. I had to fill in CUDAMEMCPYXTOY because I want to specify the stream, but the better way of course is to do STREAM=STREAM
- yes, all memcpy are memcpy(dst, [...], src, [...]). So for H2D, you have DEV first, and for D2H you have HST first. this is true for the normal C memcpy, but also for all cudaMemcpy variants.

Thanks for doing the merge with master! I merged this back with my branch and did this minor change. I also slightly change the test quoting explicitly which function is supposed to be called.